cosim/black-parrot-example/ps.cpp

```diff
@@ -135,45 +183,67 @@ extern "C" void cosim_main(char *argstr) {
 
   long val;
   bsg_pr_info("ps.cpp: reading three base registers\n");
-  bsg_pr_info("ps.cpp: reset(lo)=%ld dram_init=%ld, dram_base=%ld\n",
-              zpl->axil_read(0x0 + gp0_addr_base),
-              zpl->axil_read(0x4 + gp0_addr_base),
-              val = zpl->axil_read(0x8 + gp0_addr_base));
-
-  bsg_pr_info("ps.cpp: putting BP into reset\n");
-  zpl->axil_write(0x0 + gp0_addr_base, 0x0, mask1); // BP reset
+  bsg_pr_info("ps.cpp: reset(lo)=%d dram_init=%d, dram_base=%x\n",
+              zpl->axil_read(GP0_RD_CSR_BITBANG       + gp0_addr_base),
+              zpl->axil_read(GP0_RD_CSR_DRAM_INITED + gp0_addr_base),
+              val = zpl->axil_read(GP0_RD_CSR_DRAM_BASE + gp0_addr_base));
+
+  // Reset bsg tag master
+  // This is needed because otherwise zeros_ctr_r in the tag master will never
+  //   be zero, though in hardware it should eventually be set to 0.
+
+  zpl->axil_write(GP0_WR_CSR_BITBANG + gp0_addr_base, 0x1, 0xf);
+  for(unsigned i = 0;i < get_bsg_tag_reset_len(NUM_RESET, 1);i++)
+    zpl->axil_write(GP0_WR_CSR_BITBANG + gp0_addr_base, 0x0, 0xf);
+  // Reset bsg client0
+  write_bsg_tag_packet(zpl, NUM_RESET, 1, 0, 0, -1U);
```

Comment:
is -1U not undefined behavior?

Reply:
It should be well defined. https://stackoverflow.com/questions/8391732/is-1u-valid-c

Reply:
cool!